### PR TITLE
Add analysis engine service project

### PR DESCRIPTION
**Pull request by jeremymeng**

### CODE REVIEW

File: src/backend/PortabilityService.WorkflowManagement/WorkflowActions/AnalyzeAction.cs [hunk 1]
```diff
@@ -1,17 +1,27 @@
...
         public async Task<WorkflowStage> ExecuteAsync(string submissionId, CancellationToken token)
         {
-            // TODO: Update to call Analysis Service
-            await Task.Delay(5);
+            var formContent = new FormUrlEncodedContent(new[]
+            {
+                new KeyValuePair<string, string>("submissionId", submissionId)
+            });
+
+            var response = await httpClient.PostAsync(ServiceUrl, formContent).ConfigureAwait(false);
```
[the remark this answers is not in the document]
Comment (jeremymeng) [reply src/backend/PortabilityService.WorkflowManagement/WorkflowActions/AnalyzeAction.cs]: I've push a change to mock the workflow actions to only return the next stage and not invoke service.  That particular test's goal is verify the behavior of workflow management, not the real `AnalyzeAction` anyway.

File: src/backend/PortabilityService.AnalysisEngine/appsettings.json [hunk 2]
```diff
@@ -0,0 +1,3 @@
+﻿{
+  "BlobStorageConnectionString": "tbd"
```
Comment (chlowell) [inline src/backend/PortabilityService.AnalysisEngine/appsettings.json]: Could we leave this undefined in production? We have other, more flexible configuration sources there.

File: src/backend/PortabilityService.WorkflowManagement/WorkflowActions/BaseAction.cs [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+using System;
+using System.Net.Http;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace PortabilityService.WorkflowManagement
+{
+    abstract class BaseAction
```
Comment (chlowell) [inline src/backend/PortabilityService.WorkflowManagement/WorkflowActions/BaseAction.cs]: Nitpicking: an access modifier?

File: src/lib/Microsoft.Fx.Portability.Azure/Storage/StorageBlob.cs [hunk 4]
```diff
@@ -70,14 +78,76 @@ public StorageBlob(CloudStorageAccount storageAccount)
             }
         }
 
+        public async Task SaveResultToBlobAsync(string uniqueId, AnalyzeResponse result)
+        {
+            var container = await GetBlobContainerAsync(ResultContainerNamePrefix).ConfigureAwait(false);
+
+            if (container == null)
+            {
+                throw new InvalidOperationException("Azure CloubBlobContainer is not expected to be null");
```
Comment (chlowell) [inline src/lib/Microsoft.Fx.Portability.Azure/Storage/StorageBlob.cs]: Could use a throw expression here i.e. `var container = await ... ?? throw ...`

File: src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs [hunk 5]
```diff
@@ -0,0 +1,115 @@
...
+                if (request == null)
+                {
+                    _logger.LogError("Request for {submissionId} not found", submissionId);
+                    return NotFound();
+                }
+
+                // if the user opted out of us collecting telemetry
+                if (!request.RequestFlags.HasFlag(AnalyzeRequestFlags.NoTelemetry))
+                {
+                    //TODO: remove the blob from Azure Blob Storage
```
[the remark this answers is not in the document]
Comment (jeremymeng) [reply src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs]: Good point.  We should always check for this flag when doing telemetry work on requests.

---
In reply to: [186491469](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186491469) [](ancestors = 186491469)

File: src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs [hunk 6]
```diff
@@ -0,0 +1,115 @@
...
+                var request = await _storage.RetrieveRequestAsync(submissionId);
+
+                if (request == null)
+                {
+                    _logger.LogError("Request for {submissionId} not found", submissionId);
+                    return NotFound();
+                }
+
+                // if the user opted out of us collecting telemetry
+                if (!request.RequestFlags.HasFlag(AnalyzeRequestFlags.NoTelemetry))
```
[the remark this answers is not in the document]
Comment (jeremymeng) [reply src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs]: Yes, I copied the old code but forgot to flip it.

---
In reply to: [186490499](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186490499) [](ancestors = 186490499)

File: src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs [hunk 7]
```diff
@@ -0,0 +1,115 @@
...
+                // but older clients (alpha, VS plugin) as of 3/13/2015 do not show this and will submit a request with
+                // no specified reports
+                if (analyzeRequest.RequestFlags == AnalyzeRequestFlags.None || analyzeRequest.RequestFlags == AnalyzeRequestFlags.NoTelemetry)
+                {
+                    analyzeRequest.RequestFlags |= AnalyzeRequestFlags.ShowNonPortableApis;
+                }
+
+                //TODO: invoke the real analysis engine to do the work
+                //return _requestAnalyzer.AnalyzeRequest(analyzeRequest, submissionId);
+                await Task.Yield(); // to remove the warning about missing await in an async method
```
[the remark this answers is not in the document]
Comment (jeremymeng) [reply src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs]: Yes that works better.

---
In reply to: [186497373](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186497373) [](ancestors = 186497373)

File: src/backend/PortabilityService.WorkflowManagement/WorkflowActions/BaseAction.cs (hunk 3 above)
Comment (jeremymeng) [reply src/backend/PortabilityService.WorkflowManagement/WorkflowActions/BaseAction.cs]: Fixed: `internal` added for base and derived actions.

---
In reply to: [186502322](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186502322) [](ancestors = 186502322)

File: src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs (hunk 7 above)
[the remark this answers is not in the document]
Comment (jeremymeng) [reply src/backend/PortabilityService.AnalysisEngine/Controllers/AnalyzeController.cs]: Removed

---
In reply to: [186513790](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186513790) [](ancestors = 186513790,186497373)

File: src/backend/PortabilityService.AnalysisEngine/appsettings.json (hunk 2 above)
Comment (jeremymeng) [reply src/backend/PortabilityService.AnalysisEngine/appsettings.json]: Removed

---
In reply to: [186501307](https://github.com/Microsoft/dotnet-apiport/pull/635#discussion_r186501307) [](ancestors = 186501307)

**Pull request merged**